[RFT] lantiq: add basic 6.12 kernel support

DragonBluep opened the pull request:
### Todo:
- [x] Waiting for the prerequisites to be merged.
https://github.com/openwrt/openwrt/pull/18744
https://github.com/openwrt/openwrt/pull/18768

**_Build test only. It may brick your device. Don't try it if you are not a developer._**

DragonBluep committed:
lantiq: misc build warning fixes for upstream codes

This patchset fixes a lot of compilation warnings. In the upcoming
6.12 kernel, these warnings will be treated as errors.

Link:
https://lore.kernel.org/all/OSBPR01MB1670163BDCA60924B3671D45BC72A@OSBPR01MB1670.jpnprd01.prod.outlook.com/

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>

DragonBluep committed:
lantiq: fix timer driver build warnings

This patch fixes the following build warnings:

arch/mips/lantiq/xway/timer.c:160:14: error: no previous prototype for 'ltq_get_fpi_bus_clock' [-Werror=missing-prototypes]
  160 | unsigned int ltq_get_fpi_bus_clock(int fpi) {
      |              ^~~~~~~~~~~~~~~~~~~~~
arch/mips/lantiq/xway/timer.c:803:36: error: invalid use of undefined type 'struct platform_device'
  803 |         ret = devm_mutex_init(&pdev->dev, &timer_dev.gptu_mutex);
      |                                    ^~
arch/mips/lantiq/xway/timer.c:864:12: error: no previous prototype for 'lq_gptu_init' [-Werror=missing-prototypes]
  864 | int __init lq_gptu_init(void)
      |            ^~~~~~~~~~~~
arch/mips/lantiq/xway/timer.c:873:13: error: no previous prototype for 'lq_gptu_exit' [-Werror=missing-prototypes]
  873 | void __exit lq_gptu_exit(void)
      |             ^~~~~~~~~~~~

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>

DragonBluep committed:
lantiq: fix eeprom driver build warnings:

This patch fixes the following build warnings:

arch/mips/lantiq/xway/ath5k_eep.c:34:12: error: no previous prototype for 'of_ath5k_eeprom_probe' [-Werror=missing-prototypes]
   34 | int __init of_ath5k_eeprom_probe(struct platform_device *pdev)
      |            ^~~~~~~~~~~~~~~~~~~~~
arch/mips/lantiq/xway/eth_mac.c:15:11: error: no previous prototype for 'ltq_get_eth_mac' [-Werror=missing-prototypes]
   15 | const u8* ltq_get_eth_mac(void)
      |           ^~~~~~~~~~~~~~~

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>

DragonBluep committed:
lantiq: fix pcie driver build warnings

This patch fixes the following build warnings:

arch/mips/pci/ifxmips_pcie.c:1084:36: error: invalid use of undefined type 'struct platform_device'
 1084 |     struct device_node *node = pdev->dev.of_node;
      |                                    ^~
arch/mips/pci/ifxmips_pcie.c:1204:12: error: no previous prototype for 'ifx_pcie_bios_init' [-Werror=missing-prototypes]
 1204 | int __init ifx_pcie_bios_init(void)
      |            ^~~~~~~~~~~~~~~~~~

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>

DragonBluep committed:
lantiq: fix vmmc driver build warnings

This patch fixes the following build warnings:

arch/mips/lantiq/xway/vmmc.c:31:14: error: no previous prototype for 'ltq_get_mps_ad0_irq' [-Werror=missing-prototypes]
   31 | unsigned int ltq_get_mps_ad0_irq(void)
      |              ^~~~~~~~~~~~~~~~~~~
arch/mips/lantiq/xway/vmmc.c:40:14: error: no previous prototype for 'ltq_get_mps_ad1_irq' [-Werror=missing-prototypes]
   40 | unsigned int ltq_get_mps_ad1_irq(void)
      |              ^~~~~~~~~~~~~~~~~~~
arch/mips/lantiq/xway/vmmc.c:49:14: error: no previous prototype for 'ltq_get_mps_vc_irq' [-Werror=missing-prototypes]
   49 | unsigned int ltq_get_mps_vc_irq(int idx)
      |              ^~~~~~~~~~~~~~~~~~

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>

DragonBluep committed:
lantiq: refresh kernel symbol configs

This is a preparation for introducing the 6.12 kernel support.
All configs are automatically refreshed.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>

DragonBluep committed:
lantiq: rename patchset and kernel configs to 6.12

This is a preparation for 6.12 kernel support. It can help us
track the files history by using the Git tool.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>

DragonBluep committed:
lantiq: restore kernel 6.6 config files and patches

Copy patches and kernel configs from 6.12 kernel to restore the
default 6.6 kernel support files.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>

DragonBluep committed:
lantiq: remove upstreamed patches from 6.12 kernel patchset

Remove upstreamed patches with version labels in their names.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>

DragonBluep committed:
lantiq: refresh 6.12 kernel patches

Manually rebased:
001-MIPS-lantiq-add-pcie-driver.patch
028-NET-lantiq-various-etop-fixes.patch
152-lantiq-VPE.patch
160-owrt-lantiq-multiple-flash.patch

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>

DragonBluep committed:
lantiq: lantiq-flash: platform driver .remove_new conversion

The .remove has been converted to .remove_new in the 6.12 kernel.
We have to change remove function return type from int to void
to match the upstream changes.

Link: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-6.12.y&id=5882bf98089f3a9ecbee5a57ae0aec6f54797855
Signed-off-by: Shiji Yang <yangshiji66@outlook.com>

DragonBluep committed:
lantiq: refresh 6.12 kernel config files

All kernel symbols are automatically refreshed by
`make kernel_oldconfig CONFIG_TARGET=target` and
`make kernel_oldconfig CONFIG_TARGET=subtarget`.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>

DragonBluep committed:
lantiq: enable 6.12 testing kernel

The 6.12 testing kernel for lantiq target is ready now.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>

DragonBluep committed:
lantiq: fix i2c driver compatibility issue on 6.12 kernel

* Drop obsolete flag I2C_CLASS_SPD. [1]
* Change .remove callback return type to NULL.
* Use strscpy instead of strlcpy.

This patch fixes the following build errors:

drivers/i2c/busses/i2c-lantiq.c:599:41: error: 'I2C_CLASS_SPD' undeclared (first use in this function); did you mean 'I2C_CLASS_HWMON'?
  599 |         adap->class = I2C_CLASS_HWMON | I2C_CLASS_SPD;
      |                                         ^~~~~~~~~~~~~
      |                                         I2C_CLASS_HWMON
drivers/i2c/busses/i2c-lantiq.c:600:9: error: implicit declaration of function 'strlcpy'; did you mean 'strncpy'? [-Wimplicit-function-declaration]
  600 |         strlcpy(adap->name, DRV_NAME "-adapter", sizeof(adap->name));
      |         ^~~~~~~
      |         strncpy
drivers/i2c/busses/i2c-lantiq.c:729:19: error: initialization of 'void (*)(struct platform_device *)' from incompatible pointer type 'int (*)(struct platform_device *)' [-Wincompatible-pointer-types]
  729 |         .remove = ltq_i2c_remove,
      |

[1] https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-6.12.y&id=e61bcf42d290e73025bab38e0e55a5586c2d8ad5
Signed-off-by: Shiji Yang <yangshiji66@outlook.com>

nouman8 commented:
Getting this compile error
`make[9]: *** [scripts/Makefile.build:229: /storage/6.12/openwrt/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/drv_mei_cpe-1.5.17.6/src/drv_mei_cpe_linux.o] Error 1
make[8]: *** [/storage/6.12/openwrt/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/linux-6.12.27/Makefile:1949: /storage/6.12/openwrt/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/drv_mei_cpe-1.5.17.6/src] Error 2
make[7]: *** [Makefile:224: __sub-make] Error 2
make[7]: Leaving directory '/storage/6.12/openwrt/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/linux-6.12.27'
make[6]: *** [Makefile:2260: drv_mei_cpe.ko] Error 2
make[6]: Leaving directory '/storage/6.12/openwrt/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/drv_mei_cpe-1.5.17.6/src'
make[5]: *** [Makefile:793: all] Error 2
make[5]: Leaving directory '/storage/6.12/openwrt/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/drv_mei_cpe-1.5.17.6/src'
make[4]: *** [Makefile:431: all-recursive] Error 1
make[4]: Leaving directory '/storage/6.12/openwrt/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/drv_mei_cpe-1.5.17.6'
make[3]: *** [Makefile:95: /storage/6.12/openwrt/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/drv_mei_cpe-1.5.17.6/.built] Error 2
make[3]: Leaving directory '/storage/6.12/openwrt/package/kernel/lantiq/ltq-vdsl-vr9-mei'
time: package/kernel/lantiq/ltq-vdsl-vr9-mei/compile#4.90#1.31#5.93
    ERROR: package/kernel/lantiq/ltq-vdsl-vr9-mei failed to build.
make[2]: *** [package/Makefile:189: package/kernel/lantiq/ltq-vdsl-vr9-mei/compile] Error 1
make[2]: Leaving directory '/storage/6.12/openwrt'
make[1]: *** [package/Makefile:183: /storage/6.12/openwrt/staging_dir/target-mips_24kc_musl/stamp/.package_compile] Error 2
make[1]: Leaving directory '/storage/6.12/openwrt'
make: *** [/storage/6.12/openwrt/include/toplevel.mk:233: world] Error 2`

nouman8 commented:
Compiled and Tested on BT-HH5A along with PR https://github.com/openwrt/openwrt/pull/18744
Compiled a very minimal build, so far Wifi and VDSL are working flawlessly.
Thank You @DragonBluep :)

neheb commented:
@DragonBluep should the opportunity be taken here to remove of_get_named_gpio from the pci driver?

neheb commented:
oh this was done upstream: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=90c2d2eb7ab5848c4f853751f12b96bdc460ad1b

DragonBluep committed:
lantiq: add the missing function pointer ltq_pci_plat_dev_init

Partial revert upstream commit 075dd2eef025 ("MIPS: lantiq: Remove
unused function pointer variables") to fix the build error:

mips-openwrt-linux-musl-ld: arch/mips/lantiq/xway/ath5k_eep.o: in function `of_ath5k_eeprom_probe':
arch/mips/lantiq/xway/ath5k_eep.c:(.init.text+0x31c): undefined reference to `ltq_pci_plat_dev_init'
mips-openwrt-linux-musl-ld: arch/mips/lantiq/xway/ath5k_eep.c:(.init.text+0x338): undefined reference to `ltq_pci_plat_dev_init'

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>

lantiq: fix vmmc driver build warnings

abajk commented on target/linux/lantiq/patches-6.12/102-09-net-phy-psb6970-fix-missing-prototypes-warnings.patch:
```diff
@@ -0,0 +1,38 @@
+ 	.read_status = &psb6970_read_status,
+ };
+ 
+-int __init psb6970_init(void)
```
This driver is downstream. It would be preferable to modify its sources in target/linux/generic/files/drivers/net/phy/psb6970.c.

DragonBluep replied:
Thanks. I've opened a new PR for it: https://github.com/openwrt/openwrt/pull/19174

abajk commented on target/linux/lantiq/patches-6.12/102-01-MIPS-lantiq-xway-mark-ltq_ar9_sys_hz-as-static.patch:
```diff
@@ -0,0 +1,26 @@
+ 	return clk;
+ }
+ 
+-unsigned long ltq_ar9_sys_hz(void)
```
This should go upstream. Did you consider sending the patch upstream?

DragonBluep replied:
Patch link added.
